### PR TITLE
Edge Function Middleware API

### PR DESCRIPTION
**Pull request by migs647**
Updates to analytics API

### CODE REVIEW

File: analytics-samples/analytics-sample/build.gradle [hunk 1]
```diff
@@ -46,4 +46,5 @@ dependencies {
   androidTestImplementation 'com.segment.backo:backo:1.0.0'
 
   androidTestImplementation 'org.assertj:assertj-core:3.16.1'
+  implementation project(':jsmiddleware')
```
Comment (migs647) [inline analytics-samples/analytics-sample/build.gradle:49]: This needs to be removed

File: analytics-samples/analytics-sample/src/main/java/com/segment/analytics/sample/SampleApp.java [hunk 2]
```diff
@@ -25,82 +25,99 @@
...
 import io.github.inflationx.calligraphy3.CalligraphyConfig;
 import io.github.inflationx.calligraphy3.CalligraphyInterceptor;
 import io.github.inflationx.viewpump.ViewPump;
 
 public class SampleApp extends Application {
 
-  // https://segment.com/segment-engineering/sources/android-test/settings/keys
-  private static final String ANALYTICS_WRITE_KEY = "HO63Z36e0Ufa8AAgbjDomDuKxFuUICqI";
+    // https://segment.com/segment-engineering/sources/android-test/settings/keys
+    private static final String ANALYTICS_WRITE_KEY = "1vCwH2CEddxqO0nsk7rKezv0gzFNlJGH";
```
Comment (migs647) [inline analytics-samples/analytics-sample/src/main/java/com/segment/analytics/sample/SampleApp.java:43]: Should we reset this write key?

Comment (migs647) [reply analytics-samples/analytics-sample/src/main/java/com/segment/analytics/sample/SampleApp.java:43]: Maybe we should do something generic key with a TODO above it for when people want to use the sample app they'll know where to add their own key easily via the IDE

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 3]
```diff
@@ -1310,6 +1315,16 @@ public Builder useDestinationMiddleware(String key, Middleware middleware) {
       return this;
     }
 
+    public Builder useEdgeFunctionMiddleware(JSMiddleware middleware) {
+
```
Comment (prayansh) [inline analytics/src/main/java/com/segment/analytics/Analytics.java:1319]: We should add an IllegalStateException here, in the case that sourceMiddleware/destinationMiddleware are already used

File: analytics/src/main/java/com/segment/analytics/ConnectionFactory.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
 package com.segment.analytics;
 
 import android.util.Base64;
-import com.segment.analytics.core.BuildConfig;
+import com.segment.analytics.core.*;
```
Comment (prayansh) [inline analytics/src/main/java/com/segment/analytics/ConnectionFactory.java:27]: what is this change for?

File: build.gradle [hunk 5]
```diff
@@ -1,5 +1,6 @@
 buildscript {
-  ext.kotlin_version = '1.3.72'
+    ext.kotlin_version = '1.4.0'
+    ext.kotlin_version = '1.3.72'
```
Comment (prayansh) [inline build.gradle:3]: lets delete this line

File: analytics-samples/analytics-sample/build.gradle (hunk 1 above)
Comment (prayansh) [inline analytics-samples/analytics-sample/build.gradle:49]: lets delete this line

File: settings.gradle [hunk 6]
```diff
@@ -3,3 +3,5 @@ include 'analytics-tests'
 include 'analytics-wear'
 include 'analytics-samples:analytics-sample'
 include 'analytics-samples:analytics-wear-sample'
+include 'jsruntime'
+include 'jsmiddleware'
```
Comment (prayansh) [inline settings.gradle:7]: lets delete these  lines

File: analytics/src/main/java/com/segment/analytics/Analytics.java [hunk 7]
```diff
@@ -1398,11 +1413,20 @@ public Analytics build() {
...
               ? Collections.<String, List<Middleware>>emptyMap()
               : immutableCopyOf(this.destinationMiddleware);
+      if (this.edgeFunctionMiddleware != null
+          && this.edgeFunctionMiddleware.sourceMiddleware != null) {
+        srcMiddleware = this.edgeFunctionMiddleware.sourceMiddleware;
+      }
+      if (this.edgeFunctionMiddleware != null
+          && this.edgeFunctionMiddleware.destinationMiddleware != null) {
+        destMiddleware = this.edgeFunctionMiddleware.destinationMiddleware;
+      }
```
Comment (prayansh) [inline analytics/src/main/java/com/segment/analytics/Analytics.java:1429]: I think this could raise a scenario where edge function is inter-operating with either native source/destination middleware. 
Lets try to have parity with iOS here

**Review by prayansh — COMMENTED** [review]
left some comments

**Review by prayansh — CHANGES_REQUESTED** [review]
Lets also add some tests for the new behaviours

File: analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionMiddlewareOverwrites() {
+
+        try {
+            val analytics = builder
+                    .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                    .useSourceMiddleware(
```
Comment (prayansh) [inline analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt:96]: just wanna confirm that we are ok with letting users specify both without raising an exception in this scenario

File: analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .useSourceMiddleware(
+                            Middleware { throw AssertionError("should not be invoked") })
+                    .build()
+
+            Assertions.fail("Should not reach this state")
+        } catch (exception: java.lang.Exception) {
+            Assertions.assertThat(exception).isInstanceOf(IllegalStateException::class.java)
+        }
+    }
```
Comment (prayansh) [inline analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt:140]: can we also add a test for the scenario where we add a source/destination middleware before edge functions and an exception is raised?

**Review by prayansh — APPROVED** [review]
LGTM
